Archiving segments using DSN.

shamil-gadelshin opened the pull request:
This PR adds segment archiving via DSN to farmers and nodes. Node pushishes an archived segment using the common pubsub topic from the networking crate and farmer receives this segment via gossipsub protocol. Farmer changes doesn't include processing of archived segments.

### Changes
- subspace-node
	- starts a separate dsn-archiving task with gossipsub publishing if DSN configuration was provided on startup
	- new command line parameter expected after `--` - `--dsn-listen-on`
- subspace-farmer
	- subscribes to archiving topic if DSN configuration was enabled on startup
	- new command line parameter expected `--enable-dsn-archiving`

#### Comments
- archived segments should be probably changed to pieces 
- `--dsn-listen-on` must be in multiaddress format now, however it could be changed to just a port
- DSN on the node is configured via the same `Ed25519` keypair

shamil-gadelshin committed:
networking: Update gossipsub configuration, add common pub-sub topic.

shamil-gadelshin committed:
Add DSN archiving to subspace node (update service, add command line params).

shamil-gadelshin committed:
farmer: Add DSN archiving, add configuration parameters.

i1i1 commented on crates/subspace-networking/src/lib.rs:
```diff
@@ -30,5 +30,13 @@ mod utils;
+lazy_static! {
+    /// Topic for gossipsub protocol subscription (segment archiving).
+    pub static ref PUB_SUB_ARCHIVING_TOPIC: Sha256Topic =
+        Sha256Topic::new("PUB_SUB_ARCHIVING_TOPIC");
```
I think it would be better to use [this](https://doc.rust-lang.org/std/lazy/struct.Lazy.html) from nightly, or just `once_cell` crate. What do you think @nazar-pc

shamil-gadelshin replied:
I would keep it. Those features are unstable and can be removed, if there is an alternative - it's better to not use it. And [once_cell](https://docs.rs/once_cell/latest/once_cell/) looks like having a different use-case.

i1i1 replied:
Hmm, those are literally the same. I meant using this:
https://docs.rs/once_cell/latest/once_cell/sync/struct.Lazy.html

shamil-gadelshin replied:
You're probably right. https://docs.rs/once_cell/latest/once_cell/#faq 
I could change it if we prefer `once_cell` in the project. @nazar-pc

nazar-pc replied:
`once_cell` should be preferred if possible just like FAQ there says. You typically create static variable and a function to access it with `once_cell`.

i1i1 commented on crates/subspace-service/src/dsn_archiver.rs:
```diff
@@ -0,0 +1,55 @@
+                            trace!("Archived segment published.");
+                        }
+                        Err(err) => {
+                            error!("Failed to publish archived segment: {:?}", err);
```
```rust
error!(error = ?err, "Failed to publish archived segment");
```

shamil-gadelshin replied:
Fixed here: https://github.com/subspace/subspace/pull/630/commits/4ec48d8bd2c280d909b0d7ece51eb7861677643f

i1i1 commented on crates/subspace-farmer/src/multi_farming.rs:
```diff
@@ -179,6 +182,33 @@ impl MultiFarming {
+                                }
+                            }
+                            Err(err) => {
+                                error!("Pubsub archiving subscription failed: {:?}", err);
```
Same here

shamil-gadelshin replied:
Fixed here: https://github.com/subspace/subspace/pull/630/commits/4ec48d8bd2c280d909b0d7ece51eb7861677643f

i1i1 commented on crates/subspace-farmer/src/multi_farming.rs:
```diff
@@ -179,6 +182,33 @@ impl MultiFarming {
+                                info!("Subscribed to pubsub archiving.");
+
+                                //TODO: Integrate with archiving process.
+                                #[allow(clippy::redundant_pattern_matching)]
```
Why does it complain here though?

shamil-gadelshin replied:
I didn't use the result from the `subscription.next()`. But it's a place holder for the future integration.

i1i1 reviewed (commented):
Looks good apart from some minor issues. Should we also disable syncing via subscription in the same pr?

shamil-gadelshin committed:
Refactor comments.

nazar-pc commented on crates/subspace-node/src/bin/subspace-node.rs:
```diff
@@ -366,10 +376,36 @@ fn main() -> Result<(), Error> {
+                            primary_chain_config.network.node_key.clone().into_keypair()
+                        {
+                            let args: DsnConfigurationArgs =
+                                DsnConfigurationArgs::parse_from(cli.secondary_chain_args.clone());
```
Those are arguments for secondary node, you don't need to touch them, it is for an executor exclusively. You need to add an argument to the primary node instead. And not with a hack like here, but instead to `Cli` data structure used above already. It has flattened `RunCmd`, you can add another parameter after that.

nazar-pc commented on crates/subspace-node/src/bin/subspace-node.rs:
```diff
@@ -100,8 +101,17 @@ fn force_use_parity_db(database_params: &mut DatabaseParams) {
+pub struct DsnConfigurationArgs {
+    /// DSN listen on multi-address
+    #[clap(long)]
+    pub dsn_listen_on: Option<Multiaddr>,
```
It will be moved from here, but either way, it should be `Vec<Multiaddr>`, not `Option<Multiaddr>`

nazar-pc commented on crates/subspace-service/Cargo.toml:
```diff
@@ -54,10 +54,12 @@ sp-timestamp = { version = "4.0.0-dev", git = "https://github.com/subspace/subst
 substrate-frame-rpc-system = { version = "4.0.0-dev", git = "https://github.com/subspace/substrate", rev = "5f0aa1feb7250ac7b8c1b9928f87b2420b530e22" }
 substrate-prometheus-endpoint = { git = "https://github.com/subspace/substrate", rev = "5f0aa1feb7250ac7b8c1b9928f87b2420b530e22" }
 thiserror = "1.0.31"
+tokio = { version = "1.18.2" }
```
Please no tokio here, spawn with `spawner` such that everything can be shut down when node shuts down (tokio runtime will continue working when this happens).

nazar-pc commented on crates/subspace-node/src/bin/subspace-node.rs:
```diff
@@ -366,10 +376,36 @@ fn main() -> Result<(), Error> {
+                        let mut config = None;
+
+                        if let Ok(keypair) =
+                            primary_chain_config.network.node_key.clone().into_keypair()
```
This will create Ed25519 keypair, not Sr25519. I guess it doesn't really matter for node :slightly_smiling_face:

shamil-gadelshin replied:
Yes. We discussed that earlier.

nazar-pc commented on crates/subspace-service/src/dsn_archiver.rs:
```diff
@@ -0,0 +1,55 @@
+use futures::StreamExt;
+use sc_consensus_subspace::{ArchivedSegmentNotification, SubspaceLink};
+use sp_core::traits::SpawnEssentialNamed;
+use sp_core::Encode;
```
Not sure why you need `sp_core::Encode`. Add to dependencies and use SCALE codec instead.

nazar-pc commented on crates/subspace-service/src/dsn_archiver.rs:
```diff
@@ -0,0 +1,55 @@
+                .subscribe();
+
+            async move {
+                let (node, node_runner) = subspace_networking::create(node_config).await.unwrap();
```
Zero `.unwrap()` tolerance on the node, always use `.expect()` and prove why it will never happen.

Also I think networking needs to be instantiated outside of background task. Then you can start two background tasks: one for node runner and another for handling archiver notifications.

shamil-gadelshin replied:
`unwrap()` is an artifact from the early prototype - I changed it to `match` with error logging.

shamil-gadelshin replied:
I would not separate those tasks because it would require introducing some async data channel between them which is IMHO overkill for this method.

nazar-pc replied:
I don't see why synchronization is needed between them. Runner and archiver are already running in two separate background tasks, the only difference is that if network creation fails, it will fail in background tasks and crash an app later instead of not letting an app start fully in the first place.

nazar-pc commented on crates/subspace-service/src/dsn_archiver.rs:
```diff
@@ -0,0 +1,55 @@
+) where
+    Block: BlockT,
+{
+    spawner.spawn_essential_blocking(
```
Does it need to be blocking though? Looks like there isn't anything heavy here, so we can use regular spawning without occupying a separate thread.

nazar-pc commented on crates/subspace-service/src/lib.rs:
```diff
@@ -35,6 +37,7 @@ use sc_consensus_subspace::{
     RewardSigningNotification, SubspaceLink, SubspaceParams,
 };
 use sc_executor::{NativeElseWasmExecutor, NativeExecutionDispatch};
+pub use sc_network::Keypair;
```
Why do we need this re-export? I'd prefer to not have this and below re-exports, they have a tendency of being forgotten afterwards. Better add `subspace-networking` to explicit dependencies of `subspace-node`.

nazar-pc commented on crates/subspace-farmer/src/multi_farming.rs:
```diff
@@ -179,6 +182,33 @@ impl MultiFarming {
             })
             .await?;
 
+            // Enable DSN subscription for segments archiving.
+            if enable_dsn_archiving {
```
There are two things missing in this construction:
1) Disabling existing archiving mechanism (fine to not have this until actually implemented)
2) Archiving as implemented here with background task is not possible to interrupt, but we need that

shamil-gadelshin replied:
I would rather postpone both these tasks until the actual implementation. I can add some cancellation future in the background task but I'm not sure it will be the actual interface.

nazar-pc reviewed (changes requested):
Generally this is the approach I expected :+1:

shamil-gadelshin committed:
Fix review comments.

- refactor background tasks
- change node cmd arguments
- migrate to once_cell
- minor refactoring

shamil-gadelshin committed:
Merge branch 'main' into pubsub-archiving

# Conflicts:
#	crates/subspace-farmer/src/bin/subspace-farmer/commands/bench.rs
#	crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs
#	crates/subspace-farmer/src/bin/subspace-farmer/main.rs
#	crates/subspace-farmer/src/multi_farming.rs
#	crates/subspace-node/src/bin/subspace-node.rs
#	crates/subspace-service/Cargo.toml

nazar-pc commented on crates/subspace-networking/Cargo.toml:
```diff
@@ -21,7 +21,7 @@ bytes = "1.1.0"
 futures = "0.3.21"
 hex = "0.4.3"
-lazy_static = "1.4.0"
+once_cell = "1.12.0"
```
Not sorted again :disappointed:

nazar-pc commented on crates/subspace-service/src/dsn_archiver.rs:
```diff
@@ -11,11 +12,14 @@ use tracing::{error, info, trace};
     subspace_link: &SubspaceLink<Block>,
     node_config: subspace_networking::Config,
-    spawner: &impl SpawnEssentialNamed,
+    task_manager: &TaskManager,
```
Old spawner argument was fine actually, spawner is clonable if you need that (which I don't think is actually the case here).

nazar-pc reviewed (commented):
I'll send a PR on top of this one shortly

nazar-pc committed:
Minor tweaks and some TODOs

nazar-pc committed:
Refactor DSN node in `subspace-service` into two tasks

nazar-pc reviewed (dismissed):
Pushed a few changes:
* no need to print errors when listen multiaddress is not specified for node
* added TODOs and did some minor refactoring
* changes dsn in `subspace-service` to run two tasks and create instance before returning

Feel free to merge if this makes sense to you.

liuchengxu commented on crates/subspace-service/src/subspace_dsn.rs:
```diff
@@ -0,0 +1,62 @@
+                archived_segment, ..
+            }) = archived_segment_notification_stream.next().await
+            {
+                trace!("ArchivedSegmentNotification received");
```
Probably nicer to add a target to the log like `trace!(target: "dsn", "Subspace DSN archiver started.");` to achieve some flexibility?

liuchengxu commented on crates/subspace-node/src/bin/subspace-node.rs:
```diff
@@ -355,16 +355,37 @@ fn main() -> Result<(), Error> {
+                            .into_keypair()
+                            .map_err(|error| {
+                                sc_service::Error::Other(format!(
+                                    "Failed to conver network keypair: {error:?}"
```
```suggestion
                                    "Failed to convert network keypair: {error:?}"
```

liuchengxu commented on crates/subspace-service/src/dsn.rs:
```diff
@@ -0,0 +1,62 @@
+                archived_segment, ..
+            }) = archived_segment_notification_stream.next().await
+            {
+                trace!("ArchivedSegmentNotification received");
```
Looks like the target is missed, and I'm not sure if we should also display the segment index or something like that in it as this printing without more context does not seem to be super fruitful.

nazar-pc replied:
This is mostly just a placeholder for future logic IMO, we'll likely broadcast pieces instead of segments across multiple topics.

i1i1 commented on crates/subspace-node/src/bin/subspace-node.rs:
```diff
@@ -355,34 +355,30 @@ fn main() -> Result<(), Error> {
+                                ))
+                            })?;
+
+                        (!cli.dsn_listen_on.is_empty()).then(|| subspace_networking::Config {
+                            keypair: network_keypair,
+                            listen_on: cli.dsn_listen_on,
+                            ..subspace_networking::Config::with_generated_keypair()
+                        })
```
This should be easier to read:
```rust
use std::ops::Not;

cli.dsn_listen_on.is_empty().not().then(|| ...)
```

shamil-gadelshin committed:
Merge branch 'main' into pubsub-archiving
